lib/Query_Processor.cpp

```diff
@@ -620,123 +620,219 @@ void * get_query_digests_parallel(void *_arg) {
 	return NULL;
 }
 
-void * purge_query_digests_parallel(void *_arg) {
-	get_query_digests_parallel_args *arg = (get_query_digests_parallel_args *)_arg;
-	unsigned long long i = 0;
-	unsigned long long r = 0;
-	unsigned long long m = arg->m;
-	set_thread_name("PurgeQueryDgest", GloVars.set_thread_name);
-	for (std::unordered_map<uint64_t, void *>::iterator it=arg->gu->begin(); it!=arg->gu->end(); ++it) {
-		if ((i%DIGEST_STATS_FAST_THREADS)==m) {
-			QP_query_digest_stats *qds=(QP_query_digest_stats *)it->second;
-			delete qds;
-			r++;
+struct purge_query_digests_args {
+	umap_query_digest *digest_umap;
+	umap_query_digest_text *digest_text_umap;
+	std::function<bool(std::unordered_map<uint64_t, void *>::iterator it)> delete_entry = nullptr;
+
+	// only used for synchronous multi-threaded purge operation
+	int scan_idx;
+	int digest_deleted = 0;
+};
+
+unsigned long long purge_query_digest_entry(purge_query_digests_args* args) {
+	unsigned long long digest_deleted = 0;
+
+	umap_query_digest *digest_umap = args->digest_umap;
+	umap_query_digest_text *digest_text_umap = args->digest_text_umap;
+
+	for (auto it = digest_umap->begin(); it != digest_umap->end();) {
+		// by default, delete all entries
+		bool delete_entry = true;
+
+		if (args->delete_entry) {
+			delete_entry = args->delete_entry(it);
 		}
-		i++;
-	}
-	arg->ret = r;
-	i = 0;
-	for (std::unordered_map<uint64_t, char *>::iterator it=arg->gtu->begin(); it!=arg->gtu->end(); ++it) {
-		if ((i%DIGEST_STATS_FAST_THREADS)==m) {
-			free(it->second);
+
+		if (delete_entry) {
+			auto text_it = digest_text_umap->find(it->first);
+			if (text_it != digest_text_umap->end()) {
+				free(text_it->second);
+				digest_text_umap->erase(text_it);
+			}
+
+			QP_query_digest_stats *qds = (QP_query_digest_stats *)it->second;
+			delete qds;
+
+			it = digest_umap->erase(it);
+			digest_deleted++;
+		} else {
+			it++;
 		}
 	}
-	return NULL;
+
+	return digest_deleted;
+}
+
+
+void * purge_query_digests_parallel(void *_args) {
+	set_thread_name("PurgeQueryDigest", GloVars.set_thread_name);
```

Comment:
This will immediately assert when attempting to execute. This name exceeds the maximum name allowed for threads, triggering this assert:
```
	int rc;
	rc = pthread_setname_np(pthread_self(), name);
	assert(!rc);
```
The name should be reverted to the previous one `PurgeQueryDgest`. For executing the test sequence suggested in my concurrent purging comment, this needs to be fixed first.

Reply:
Addressed.